### ISSUE TITLE
Stutters and slow loading times

### ISSUE THREAD

**Issue by TheAwesomeGem** [TheLimePixel/Modern-Skyblock-3#8]
It takes 15 minutes to load this pack and 5 minutes to load the world. Previously with Modern Skyblock 2, it took 1 mins and 20 seconds to load the pack and 5 seconds to load the world. Also the game stutters every 4-5 seconds, especially when I look around. I go from 200 FPS(refresh rate) to 0 FPS everytime I move around my island. Here are my specs:

Loading Times for each mod: https://pastebin.com/Kghnf9kS

CPU: AMD FX 9590 (8 Cores) (5.1 GHZ OC'ed)
GPU: EVGA GTX 1080 Ti (11 GB)
RAM: 64 GB
SSD: Samsung 850 EVO 500GB
OS: Windows 10 64 Bit

**Comment by TheAwesomeGem** [comment]
Also I added Optifine.

**Comment by andremiller** [comment]
I had the same issue until I increased my memory allocation to 8.5GB (in the Twitch launcher profile). What is your memory allocation set to?
PS: I do not use OptiFine, I had the stutters without it.

**Comment by TheAwesomeGem** [comment]
I set mine to 8GB. Removing Optifine doesn't help either.

**Comment by TheAwesomeGem** [comment]
So I updated to the new version, removed Optifine, replaced BedBugs with BedPatch, turned on Vsync and now I don't get any stutter and smooth gameplay. However loading times are still long due to JEI and TexFix taking most of the loading times.

**Comment by cauliflower69** [comment]
Just to add to this. I had a similar issue with my FPS jumping up and down from 0. BetterFPS fixed it

**Comment by Delerium** [comment]
Likewise I also had these same problems.  I usually keep my minecraft at 6 Gb allocated, but this pack would freeze every 2-3 seconds at that.  When I increased to 8 Gb, it cleared up, no fps issues.  There are still very long load times though, which is a bummer, but I can deal.  If this can't be fixed, might want to change the requirements on the project page.

**Comment by Zuckerwasser** [comment]
same here allocated 6GB, freeze after every 5 seconds hold on for 2 seconds. Unplayable :(
Fps drops from 60 (limit) to 8fps

GTX1070

**Comment by asiekierka** [comment]
Well, with 270 mods, 6000 blocks and 12000 items, you shouldn't expect good loading performance.

Especially when some mod in your pack forces texture loading and model baking to happen *twice*. There's really no reason for it to do so, it greatly increases your pack's loading time memory usage (as the first pass's models are only wiped after the second pass is done) and loading time itself.

I'm not going to figure out which mod does it, there's 270 of them, but you probably should.

**Comment by jaredlll08** [comment]
When I was looking into why WaitingTime was making the pack take forever to load, I found that with WaitingTime and optifine alone, the game takes 47 seconds to loads, whereas with just WaitingTime it takes 7 seconds. Optifine definitely affects load time

**Comment by joakime** [comment]
@Zuckerwasser 6GB is not enough, the Java VM enters into many multi second stop-the-world GC events with that little memory.   At 8.5GB max allocated to the game (pointed out by @andremiller) the startup takes ~5 minutes on my 3 year old laptop.

**Comment by asiekierka** [comment]
Yes, a lot of the reason for 6GB not being enough is "some mod in your pack forces texture loading and model baking to happen twice." - it should run fine with about 4GB allocated once that is fixed and FoamFix updated (0.9.x saves more RAM).

**Comment by asiekierka** [comment]
Looking into it now.

**Comment by asiekierka** [comment]
First find: If you want to reduce the pack loading time and RAM usage during loading massively, all you have to do is drop Fence Overhaul - see [the following issue](https://minecraft.curseforge.com/projects/fence-overhaul/issues/6).

I'm not done yet, though.

**Comment by TheAwesomeGem** [comment]
Here are the loading times for each mod: 
https://pastebin.com/Kghnf9kS

Fence Overhaul takes around 5 minutes to load as @asiekierka suggested.
Texfix takes almost 2 minutes to load.

**Comment by asiekierka** [comment]
TexFix shouldn't be providing large memory usage gains for anyone using default 16x textures. In my opinion, with the added performance load (to itself and other mods which access texture data on the CPU), it should be an optional. However, do keep in mind I am slightly biased.

As for the insane RAM usage (3.75GB with latest FoamFix and dynamicItemModels re-enabled), it is most likely tied to that issue in Fence Overhaul.

**Comment by TheAwesomeGem** [comment]
After removing both TexFix and Fence Overhaul, it now takes 3-4 minutes to load the pack rather than 15-20 mins before. That's more than 50% improvement.

**Comment by asiekierka** [comment]
I have identified the following causes:

* Second resource refresh causes there to be two ModelLoader instances present, each with their own model cache. This is caused by Fence Overhaul in this pack, but can also be caused by literally anything triggering a resource reload - so even normal user behaviour such as adding a resource pack in-game. The issue stems from mods not clearing their IModel references correctly, and some of them (VanillaModelWrapper) hold a ModelLoader reference. One such example is AE2, as reported [here](https://github.com/AppliedEnergistics/Applied-Energistics-2/issues/3370). Said second instance of ModelLoader takes up 1.2GB in memory (!), so I think it is worth looking into.
* JEI's Anvil recipe wrapper consumes almost 700MB of RAM on this pack, with 54550 instances. It appears that the pack contains a very large amount of tools - and, as such, Anvil recipes - and the amount of ItemStacks generated (each one of Silent's Gems stacks seems to take up almost 5KB!) consumes a lot of RAM. It's an inefficiency in JEI, however, in my opinion. Calling @mezz to see if something can be done about it.
* TCon's models consume a pretty impressive 620MB of RAM on this pack. I'm not sure how much I can do about this, short of recommending that Tinker's Construct developers move to using a cache (such as Guava's) instead of pre-baking all their models. @bonii-xx

Everything else seems fairly ordinary. Solving all of these should drop the modpack's RAM usage very, *very* significantly.

(Also, don't forget to re-enable FoamFix's dynamicItemModels. Drops RAM usage by 350MB.)

**Comment by asiekierka** [comment]
I don't feel TexFix contributes significantly to the loading times, I think it's more collateral damage caused by the massively increased GC load of having two ModelLoader instances present in memory during loading, as TexFix loads *after* that. Could you compare the patch with and without TexFix, but without Fence Overhaul, @TheAwesomeGem ? I don't think it should be contributing 25% to the loading time.

**Comment by TheAwesomeGem** [comment]
It still takes almost 2 minutes to load.

**Comment by asiekierka** [comment]
Huh. Okay then. Guess my hunch on this was wrong.

**Comment by TheAwesomeGem** [comment]
I don't know how to check why Texfix takes so long to load.

**Comment by bonii-xx** [comment]
>TCon's models consume a pretty impressive 620MB of RAM on this pack. I'm not sure how much I can do about this, short of recommending that Tinker's Construct developers move to using a cache (such as Guava's) instead of pre-baking all their models. @bonii-xx

What is using up that much memory? TiC already caches quite aggressively during runtime.

**Comment by asiekierka** [comment]
@bonii-xx I already deleted the heap dump, unfortunately - I'll generate a new one. You can find me on #charset on EsperNet if you'd like to discuss it, I should have one by the time you're there.

(I think part of it might've been the doubled-up ModelLoaders courtesy of Fence Overhaul, so sorry for not noticing that earlier.)

**Comment by asiekierka** [comment]
@bonii-xx Okay, so:

* The actual memory usage is 310MB, not 620MB. Apologies - the ModelLoader was doubled up, which caused it to... be doubled up.
* That's still a lot, though! Let's look into it.

![](https://img.asie.pl/snQc)

Most of that memory is used up by BakedSimpleItem. I know why that is and am looking into it now.

Keep in mind this pack includes PlusTiC - TCon on its own uses I think 1/4th of this?

**Comment by asiekierka** [comment]
@bonii-xx Is BakedSimpleItem still necessary? I don't see anything it does which recent Forge versions don't do in their own BakedItemModel (which is now a public class!), and FoamFix is capable of replacing the latter with a far more RAM-friendly implementation (I'll add special compatibility if you still need to create your own BakedItemModels by hand).

This would effectively fix most of the memory usage.

**Comment by asiekierka** [comment]
Actually... I don't think I can, as you combine various item models into one BakedSimpleItem, don't you?

**Comment by mezz** [comment]
@asiekierka here is the context on `BakedSimpleItem` and `BakedItemModel`: https://github.com/SlimeKnights/TinkersConstruct/issues/3176
I made the fix in Tinker's using `BakedSimpleItem` and then found the same issue with Forge's bucket so I fixed it there with `BakedItemModel` and made it public so it can be used by Tinker's and others.

**Comment by bonii-xx** [comment]
Oh yes, that was a really dumb bug. Iirc it wasn't like that in 1.8 and ofc the model code hardly changed since then and didn't adapt. Anyway, I think we can use the forge variants? We combine them yes, but as long es the API is the same we can use the Forge class? (Also I think the combined ones are a separate model.. and only generated at runtime on demand + cached)

**Comment by TSM-Evo2** [comment]
Also having major performance issues with the pack. 15 minute load times. Currently on 6GB allocated. I will try 8.5+ when the pack crashes again.

**Comment by ghost** [comment]
The weird thing is, I leave MC on the title screen in the background when I'm not playing, so I don't have to go through the reload (especially since I get random load crashes from Extra Bit Manipulation). Even idling in the background it causes lag spikes in the system.
